rp2040-hal/src/dma/mod.rs

```diff
@@ -0,0 +1,265 @@
+    ) => {
+        impl DMAExt for DMA {
+            fn split(self, resets: &mut pac::RESETS) -> Channels {
+                self.reset_bring_up(resets);
+
+                Channels {
+                    $(
+                        $chX: Channel {
+                            _phantom: PhantomData,
+                        },
+                    )+
+                }
+            }
+
+            fn dyn_split(self, resets: &mut pac::RESETS) -> DynChannels{
+                self.reset_bring_up(resets);
+
+                DynChannels {
+                    $(
+                        $chX: Some(Channel {
+                            _phantom: PhantomData,
+                        }),
+                    )+
+                }
+            }
+        }
+
+        /// Set of DMA channels.
+        pub struct Channels {
+            $(
+                /// DMA channel.
+                pub $chX: Channel<$CHX>,
+            )+
+        }
+        $(
+            /// DMA channel identifier.
+            pub struct $CHX;
+            impl ChannelIndex for $CHX {
+                fn id() -> u8 {
+                    $x
+                }
+            }
+        )+
+
+        /// Set of DMA channels.
+        pub struct DynChannels {
+            $(
+                /// DMA channel.
+                pub $chX: Option<Channel<$CHX>>,
+            )+
+        }
+    }
+}
+
+channels! {
+    CH0: (ch0, 0),
+    CH1: (ch1, 1),
+    CH2: (ch2, 2),
+    CH3: (ch3, 3),
+    CH4: (ch4, 4),
+    CH5: (ch5, 5),
+    CH6: (ch6, 6),
+    CH7: (ch7, 7),
+    CH8: (ch8, 8),
+    CH9: (ch9, 9),
+    CH10:(ch10, 10),
+    CH11:(ch11, 11),
+}
+
+trait ChannelRegs {
+    unsafe fn ptr() -> *const rp2040_pac::dma::CH;
+    fn regs(&self) -> &rp2040_pac::dma::CH;
+}
+
+impl<CH: ChannelIndex> ChannelRegs for Channel<CH> {
+    unsafe fn ptr() -> *const rp2040_pac::dma::CH {
+        &(*rp2040_pac::DMA::ptr()).ch[CH::id() as usize] as *const _
+    }
+
+    fn regs(&self) -> &rp2040_pac::dma::CH {
+        unsafe { &*Self::ptr() }
+    }
+}
+
+/// Trait which is implemented by anything that can be read via DMA.
+pub trait ReadTarget {
+    /// Type which is transferred in a single DMA transfer.
+    type ReceivedWord;
+
+    /// Returns the DREQ number for this data source (`None` for memory buffers).
+    fn rx_treq() -> Option<u8>;
+
+    /// Returns the address and the maximum number of words that can be transferred from this data
+    /// source in a single DMA operation.
+    ///
+    /// For peripherals, the count should likely be u32::MAX. If a data source implements
+    /// EndlessReadTarget, it is suitable for infinite transfers from or to ring buffers. Note that
+    /// ring buffers designated for endless transfers, but with a finite buffer size, should return
+    /// the size of their individual buffers here.
+    ///
+    /// # Safety
+    ///
+    /// This function has the same safety guarantees as `StaticReadBuffer::static_read_buffer`.
```

Comment:
`StaticReadBuffer` doesn't exist in the latest version of embedded_dma. Should probably be changed to `ReadBuffer::read_buffer`? At least that's how I understand the [changelog](https://github.com/rust-embedded/embedded-dma/blob/master/CHANGELOG.md): "Replace less strict ReadBuffer and WriteBuffer definitions with those of StaticReadBuffer and StaticWriteBuffer. This removes the separate static traits."

Reply:
That makes sense with my understanding of the 0.2 changes.
There's a bunch more documentation to fix up as well, unfortunately. Hoping that we'll get more interest for that with the code in-tree than we have with PRs.